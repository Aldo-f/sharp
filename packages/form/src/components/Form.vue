<template>
    <div class="SharpForm" data-popover-boundary>
        <template v-if="ready">
            <div class="container">
                <div v-show="hasErrors" class="SharpNotification SharpNotification--error" role="alert">
                    <div class="SharpNotification__details">
                        <div class="SharpNotification__text-wrapper">
                            <p class="SharpNotification__title">{{ l('form.validation_error.title') }}</p>
                            <p class="SharpNotification__subtitle">{{ l('form.validation_error.description') }}</p>
                        </div>
                    </div>
                </div>
                <TabbedLayout :layout="layout" ref="tabbedLayout">
                    <template slot="nav-prepend">
                        <template v-if="localized">
                            <LocaleSelect :locale="currentLocale" :locales="locales" @change="handleLocaleChanged" />
                        </template>
                    </template>
                    <!-- Tab -->
                    <template slot-scope="tab">
                        <Grid :rows="[tab.columns]" ref="columnsGrid">
                            <!-- column -->
                            <template slot-scope="column">
                                <FieldsLayout v-if="fields" :layout="column.fields" :visible="fieldVisible" ref="fieldLayout">
                                    <!-- field -->
                                    <template slot-scope="fieldLayout">
                                        <FieldDisplay
                                            :field-key="fieldLayout.key"
                                            :context-fields="transformedFields"
                                            :context-data="data"
                                            :field-layout="fieldLayout"
                                            :locale="fieldLocale[fieldLayout.key]"
                                            :error-identifier="fieldLayout.key"
                                            :config-identifier="fieldLayout.key"
                                            :update-data="updateData"
                                            :update-visibility="updateVisibility"
                                            @locale-change="updateLocale"
                                            ref="field"
                                        />
                                    </template>
                                </FieldsLayout>
                            </template>
                        </Grid>
                    </template>
                </TabbedLayout>
            </div>
        </template>
    </div>
</template>

<script>
    import {
        BASE_URL,
        getBackUrl, 
        getListBackUrl, 
        logError,
    } from "sharp";

    import { TabbedLayout, Grid, Dropdown, DropdownItem, } from 'sharp-ui';
    import { ActionEvents, ReadOnlyFields, Localization, DynamicView } from 'sharp/mixins';
    
    import FieldsLayout from './ui/FieldsLayout';
    import LocaleSelect from './ui/LocaleSelect';
    import localize from '../mixins/localize/form';

    import { getDependantFieldsResetData, transformFields } from "../util";
    

    const noop = ()=>{};

    export default {
        name:'SharpForm',
        extends: DynamicView,

        mixins: [ActionEvents, ReadOnlyFields('fields'), Localization, localize('fields')],

        components: {
            TabbedLayout,
            FieldsLayout,
            Grid,
            Dropdown,
            DropdownItem,
            LocaleSelect,
        },


        props:{
            entityKey: String,
            instanceId: String,

            /// Extras props for customization
            independant: {
                type:Boolean,
                default: false
            },
            ignoreAuthorizations: Boolean,
            props: Object
        },

        inject:['actionsBus'],

        provide() {
            return {
                $form:this
            }
        },

        data() {
            return {
                ready: false,

                fields: null,
                authorizations: null,
                breadcrumb: null,
                config: null,

                errors:{},
                fieldLocale: {},
                locales: null,

                fieldVisible: {},
                curFieldsetId:0,

                pendingJobs: []
            }
        },
        computed: {
            apiPath() {
                let path = `form/${this.entityKey}`;
                if(this.instanceId) path+=`/${this.instanceId}`;
                return path;
            },
            localized() {
                return Array.isArray(this.locales) && !!this.locales.length;
            },
            isSingle() {
                return this.config
                    ? this.config.isSingle
                    : false
            },
            isCreation() {
                return !this.isSingle && !this.instanceId;
            },
            isReadOnly() {
                if(this.ignoreAuthorizations) {
                    return false;
                }
                return this.isCreation
                    ? !this.authorizations.create
                    : !this.authorizations.update;
            },
            // don't show loading on creation
            synchronous() {
                return this.independant;
            },
            hasErrors() {
                return Object.keys(this.errors).some(errorKey => !this.errors[errorKey].cleared);
            },

            baseEntityKey() {
                return this.entityKey.split(':')[0];
            },

            downloadLinkBase() {
                return `/download/${this.entityKey}/${this.instanceId}`;
            },
            listUrl() {
                return `${BASE_URL}/list/${this.baseEntityKey}?restore-context=1`;
            },
            transformedFields() {
                const fields = this.isReadOnly
                    ? this.readOnlyFields
                    : this.fields;
                return transformFields(fields, this.data);
            },

            currentLocale() {
                const locales = [...new Set(Object.values(this.fieldLocale))];
                return locales.length === 1 ? locales[0] : null;
            }
        },
        methods: {
            async updateData(key, value, { forced } = {}) {
                this.data = {
                    ...this.data,
                    ...(!forced ? getDependantFieldsResetData(this.fields, key,
                        field => this.fieldLocalizedValue(field.key, null),
                    ) : null),
                    [key]: this.fieldLocalizedValue(key, value),
                }
            },
            updateVisibility(key, visibility) {
                this.$set(this.fieldVisible, key, visibility);
            },
            updateLocale(key, locale) {
                this.$set(this.fieldLocale, key, locale);
            },
            handleLocaleChanged(locale) {
                this.fieldLocale = this.defaultFieldLocaleMap({ fields: this.fields, locales: this.locales }, locale);
            },
            mount({ fields, layout, data={}, authorizations={}, locales, breadcrumb, config }) {
                this.fields = fields;
                this.data = data;
                this.layout = this.patchLayout(layout);
                this.locales = locales;
                this.authorizations = authorizations;
                this.breadcrumb = breadcrumb;
                this.config = config;

                if(fields) {
                    this.fieldVisible = Object.keys(this.fields).reduce((res, fKey) => {
                        res[fKey] = true;
                        return res;
                    },{});
                    this.fieldLocale = this.defaultFieldLocaleMap({ fields, locales });
                }
                this.validate();
            },
            validate() {
                const localizedFields = Object.keys(this.fieldLocale);
                const alert = text => this.actionsBus.$emit('showMainModal', {
                    title: 'Data error',
                    text,
                    isError: true,
                    okCloseOnly: true,
                });
                if(localizedFields.length > 0 && !this.locales.length) {
                    alert("Some fields are localized but the form hasn't any locales configured");
                }
            },
            handleError({response}) {
                if(response.status===422)
                    this.errors = response.data.errors || {};
            },

            patchLayout(layout) {
                if(!layout)return;
                let curFieldsetId = 0;
                let mapFields = layout => {
                    if(layout.legend)
                        layout.id = `${curFieldsetId++}#${layout.legend}`;
                    else if(layout.fields)
                        layout.fields.forEach(row => {
                            row.forEach(mapFields);
                        });
                };
                layout.tabs.forEach(t => t.columns.forEach(mapFields));
                return layout;
            },

            async init() {
                if(this.independant) {
                    this.mount(this.props);
                    this.ready = true;
                }
                else {
                    if(this.entityKey) {
                        await this.get();
                        this.setupActionBar();
                        this.ready = true;
                    }
                    else logError('no entity key provided');
                }
            },

            setupActionBar() {
                this.actionsBus.$emit('setup', {
                    showSubmitButton: this.isCreation
                        ? this.authorizations.create
                        : this.authorizations.update,
                    showDeleteButton: !this.isCreation && !this.isSingle && this.authorizations.delete,
                    showBackButton: this.isReadOnly,
                    opType: this.isCreation ? 'create' : 'update'
                });
            },
            redirectToList() {
                location.href = getListBackUrl(this.breadcrumb);
            },
            redirectToParentPage() {
                location.href = getBackUrl(this.breadcrumb);
            },
            async submit({ postFn }={}) {
                if(this.pendingJobs.length) {
                    return;
                }
                try {
                    const response = postFn ? await postFn(this.data) : await this.post();
                    if(this.independant) {
                        this.$emit('submit', response);
                        return response;
                    }
                    else if(response.data.ok) {
                        this.mainLoading.$emit('show');
                        this.redirectToParentPage();
                    }
                }
                catch(error) {
                    this.handleError(error);
                    return Promise.reject(error);
                }
            }
        },
        actions: {
            submit() {
                this.submit().catch(()=>{});
            },
            async 'delete'() {
                try {
                    await this.axiosInstance.delete(this.apiPath);
                    this.redirectToList();
                }
                catch(error) {

                }
            },
            cancel() {
                this.redirectToParentPage();
            },

            setPendingJob({ key, origin, value:isPending }) {
                if(isPending)
                    this.pendingJobs.push(key);
                else
                    this.pendingJobs = this.pendingJobs.filter(jobKey => jobKey !== key);

                if(this.pendingJobs.length) {
                    this.actionsBus.$emit('updateActionsState', {
                        state: 'pending',
                        modifier: origin
                    })
                }
                else {
                    this.actionsBus.$emit('updateActionsState', null);
                }
            }
        },
        created() {
            this.$on('error-cleared', errorId => {
                if(this.errors[errorId])
                    this.$set(this.errors[errorId],'cleared',true);
            })
        },
        mounted() {
            this.init();
        }
    }
</script>