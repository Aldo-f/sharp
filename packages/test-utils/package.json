{
    "name": "sharp-test-utils",
    "private": true
}