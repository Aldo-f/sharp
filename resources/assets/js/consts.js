export let API_PATH = '/sharp/api';
export let UPLOAD_URL = `${API_PATH}/upload`;