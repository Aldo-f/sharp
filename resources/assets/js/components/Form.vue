<template>
    <div class="Form container">
        <template v-if="ready">
            <template v-if="layout.tabbed && layout.tabs.length>1">
                <b-tabs pills v-model="tabIndex">
                    <b-tab v-for="(tab,i) in layout.tabs" :title="tab.title" :key="i">
                        <sharp-form-tab-content
                                :columns="tab.columns"
                                :fields="fields"
                                :data="data"
                                :update-data="updateData"
                        >
                        </sharp-form-tab-content>
                    </b-tab>
                </b-tabs>
            </template>
            <template v-else>
                <div v-for="tab in layout.tabs">
                    <sharp-form-tab-content
                            :columns="tab.columns"
                            :fields="fields"
                            :data="data"
                            :update-data="updateData"
                    >
                    </sharp-form-tab-content>
                </div>
            </template>
        </template>
        <template v-else>
            Form loading...
        </template>
    </div>
</template>

<script>
    import util from '../util';
    import TemplateDefinition from '../template-definition';
    import { API_PATH } from '../consts';
    import testableForm from '../mixins/testable-form';

    import { NameAssociation as fieldCompNames } from './fields/index';

    import FormTabContent from './FormTabContent';
    import bTabs from './vendor/bootstrap-vue/components/tabs'
    import bTab from './vendor/bootstrap-vue/components/tab'


    export default {
        name:'SharpForm',

        mixins: [
            testableForm
        ],

        components: {
            [FormTabContent.name]: FormTabContent,
            bTab, bTabs
        },

        props:{
            entityKey: String,
            instanceId: String,
        },

        data() {
            return {
                fields: null,
                data: null,
                layout: null,
                tabIndex:0,
                ready:false
            }
        },
        computed: {
            displayableFields() {
                return this.fields.filter((field, i) => {
                    if(!(field.type in fieldCompNames))
                        return util.error(`Field '${field.key}' have a unknown type (${field.type})`), false;
                    return true;
                })
            },
            apiPath() {
                let path = `${API_PATH}/form/${this.entityKey}`;
                if(this.instanceId) path+=`/${this.instanceId}`;
                return path;
            }
        },
        methods: {
            updateData(key,value) {
                this.data[key] = value;
            },
            getForm() {
                return axios.get(this.apiPath)
                    .then(({data: {fields, layout, data}}) => {
                        this.fields = fields;
                        this.layout = layout;
                        this.data = data;

                        this.ready=true;
                    });
            },
            postForm() {
                return axios.post(this.apiPath)
                    .then(response => {

                    });
            },
        },
        created() {
            if(this.entityKey != null) {
                this.getForm();
            }
            else util.error('no entity key provided');
            window.form = this;
        }
    }
</script>