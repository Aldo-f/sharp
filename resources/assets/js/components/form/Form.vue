<template>
    <div class="SharpForm">
        <template v-if="ready">
            <div v-show="hasErrors" class="SharpNotification SharpNotification__inline SharpNotification__inline--error" role="alert">
                <div class="SharpNotification__details">
                    <div class="SharpNotification__text-wrapper">
                        <p class="SharpNotification__title">{{ l('form.validation_error.title') }}</p>
                        <p class="SharpNotification__subtitle">{{ l('form.validation_error.description') }}</p>
                    </div>
                </div>
            </div>
            <sharp-tabbed-layout :layout="layout">
                <!-- Tab -->
                <template scope="tab">
                    <sharp-grid :rows="[tab.columns]">
                        <!-- column -->
                        <template scope="column">
                            <sharp-fields-layout v-if="fields" :layout="column.fields">
                                <!-- field -->
                                <template scope="fieldLayout">
                                    <sharp-field-display :field-key="fieldLayout.key"
                                                         :context-fields="isReadOnly ? readOnlyFields : fields"
                                                         :context-data="data"
                                                         :field-layout="fieldLayout"
                                                         :locale="locale"
                                                         :error-identifier="fieldLayout.key"
                                                         :update-data="updateData">
                                    </sharp-field-display>
                                </template>
                            </sharp-fields-layout>
                        </template>
                    </sharp-grid>
                </template>
            </sharp-tabbed-layout>
        </template>
    </div>
</template>

<script>
    import * as util from '../../util';
    import { API_PATH } from '../../consts';

    import { testableForm, ActionEvents, ReadOnlyFields, Localization } from '../../mixins';

    import DynamicView from '../DynamicViewMixin';

    import TabbedLayout from '../TabbedLayout'
    import Grid from '../Grid';
    import FieldsLayout from './FieldsLayout.vue';


    import axios from 'axios';

    const noop = ()=>{}

    export default {
        name:'SharpForm',
        extends: DynamicView,

        mixins: [testableForm, ActionEvents, ReadOnlyFields('fields'), Localization],

        components: {
            [TabbedLayout.name]: TabbedLayout,
            [FieldsLayout.name]: FieldsLayout,
            [Grid.name]: Grid,
        },

        props:{
            entityKey: String,
            instanceId: String,

            /// Extras props for customization
            independant: {
                type:Boolean,
                default: false
            },
            ignoreAuthorizations: Boolean,
            props: Object
        },

        inject:['actionsBus', ...DynamicView.inject],

        provide() {
            return {
                $form:this
            }
        },

        data() {
            return {
                fields: null,
                config: null,
                authorizations: null,

                errors:{},
                locale: ''
            }
        },
        computed: {
            apiPath() {
                let path = `${API_PATH}/form/${this.entityKey}`;
                if(this.instanceId) path+=`/${this.instanceId}`;
                return path;
            },
            localized() {
                return this.config && Array.isArray(this.config.locales);
            },
            isCreation() {
                return !this.instanceId;
            },
            isReadOnly() {
                return !this.ignoreAuthorizations && !(this.isCreation ? this.authorizations.create : this.authorizations.update);
            },
            // don't show loading on creation
            synchronous() {
                return this.independant;
            },
            hasErrors() {
                return Object.keys(this.errors).some(errorKey => !this.errors[errorKey].cleared);
            }
        },
        methods: {
            fieldErrors(key) {
                if(this.fields[key].localized) {
                    return (this.errors[key]||{})[this.locale];
                }
                return this.errors[key];
            },
            updateData(key,value) {
                if(this.fields[key].localized) {
                    this.$set(this.data[key],this.locale,value);
                }
                else this.$set(this.data,key,value);
            },
            mount({fields, layout, data={}, config={}, authorizations={}}) {
                this.fields = fields;
                this.layout = layout;
                this.data = data;
                this.config = config;
                this.authorizations = authorizations;
            },
            handleError({response}) {
                if(response.status===422)
                    this.errors = response.data || {};
            },
            delete() {
                axios.delete(this.apiPath);
            },
            init() {
                if(this.independant) {
                    this.mount(this.props);
                    this.ready = true;
                }
                else {
                    if(this.entityKey) {
                        this.get().then(_=>this.setupActionBar());
                    }
                    else util.error('no entity key provided');
                }
            },

            setupActionBar() {
                const showSubmitButton = this.isCreation ? this.authorizations.create : this.authorizations.update;

                this.actionsBus.$emit('setup', {
                    locales: null, //this.config.locales,
                    showSubmitButton,
                    showDeleteButton: this.authorizations.delete,
                    showBackButton: this.isReadOnly,
                    opType: this.isCreation ? 'create' : 'update'
                });

                if(this.config.locales) {
                    this.actionsBus.$emit('localeChanged', this.config.locales[0]);
                }
            },
        },
        actions: {
            submit({entityKey, endpoint, dataFormatter=noop }={}) {
                if(entityKey && entityKey !== this.entityKey) return;

                this.post(endpoint, dataFormatter(this))
                    .then(({ data })=>{
                        if(this.independant) {
                            this.$emit('submitted', data);
                        }
                        else if(data.ok) {
                            this.mainLoading.$emit('show');
                            location.href = `/sharp/list/${this.entityKey}?restore-context=1`
                        }
                    })
                    .catch(this.handleError)
            },
            cancel() {
                location.href = `/sharp/list/${this.entityKey}?restore-context=1`;
            },
            localeChanged(newLocale) {
                this.locale = newLocale;
            },
            delete: 'delete',
            reset({ entityKey }) {
                if(entityKey && entityKey !== this.entityKey) return;

                this.data = {}
            }
        },
        created() {
            this.$on('error-cleared', errorId => {
                this.$set(this.errors[errorId],'cleared',true);
            })
        },
        mounted() {
            this.init();
            console.log(this);
        }
    }
</script>