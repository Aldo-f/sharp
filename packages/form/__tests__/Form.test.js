import Vue from 'vue';
import axios from 'axios';
import * as consts from 'sharp/consts';
import Form from '../src/components/Form.vue';


import { wait, MockInjections, MockI18n, nextRequestFulfilled } from "sharp-test-utils";
import moxios from 'moxios';
import { shallowMount, createLocalVue } from '@vue/test-utils';

jest.mock('sharp');


describe('sharp-form', ()=>{
    Vue.use(MockI18n);

    function createWrapper({ propsData } = {}) {
        return shallowMount(Form, {
            propsData: {
                entityKey: 'spaceship',
                instanceId: null,
                ignoreAuthorizations: null,
                ...propsData,
            },
            provide: {
                axiosInstance: axios.create(),
                actionsBus: new Vue(),
                mainLoading: new Vue(),
            },
            stubs: {
                Grid: 
                    `<div id="MOCKED_GRID">
                        <slot v-bind="rows[0][0]" />
                    </div>`,
                FieldsLayout: 
                    `<div id="MOCKED_FIELDS_LAYOUT"> 
                        <slot v-if="layout[0][0]" v-bind="layout[0][0]" />
                    </div>`,
                FieldDisplay: true,
                TabbedLayout: 
                    `<div id="MOCKED_TABBED_LAYOUT">
                        <slot v-bind="layout.tabs[0]" />
                    </div>`,
            },
            created() {
                moxios.install(this.axiosInstance);
                moxios.uninstall = moxios.uninstall.bind(moxios, this.axiosInstance);
            }
        });
    }

    function createLayout(fields) {
        return {
            tabs : [
                {
                    columns: [
                        {
                            fields,
                        }
                    ]
                }
            ]
        }
    }

    function createForm({ key='title', type='text', ...props } = {}) {
        return {
            data: { [key]: null },
            fields: {
                [key]: {
                    key,
                    type,
                    ...props,
                }
            },
            layout: createLayout([[{ key }]]),
        }
    }


    const oldDelay = moxios.delay;
    moxios.delay = 10;

    afterAll(() => {
        moxios.delay = oldDelay;
    });
    // beforeEach(() => {
    //     moxios.install();
    // });
    afterEach(()=>{
        moxios.uninstall();
    });

    test('can mount sharp-form', async ()=>{
        const wrapper = createWrapper();

        await nextRequestFulfilled({
            status: 200,
            response: createForm(),
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    test('can mount "independant" sharp-form', ()=>{
        const wrapper = createWrapper({
            propsData:{
                independant: true,
                props: createForm()
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    test('can mount sharp-form with alert', ()=>{
        const wrapper = createWrapper({
            propsData:{
                independant: true,
                props: createForm()
            }
        });

        wrapper.setData({
            errors: {
                title: ['required']
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    test('localized', ()=>{
        const wrapper = createWrapper();
        expect(wrapper.vm.localized).toBe(false);
        wrapper.setData({
            locales: ['fr', 'en']
        });
        expect(wrapper.vm.localized).toBe(true);
    });

    test('detect when is creation', ()=>{
        const wrapper = createWrapper();

        expect(wrapper.vm.isCreation).toBe(true);

        wrapper.setProps({
            instanceId: '10',
        });

        expect(wrapper.vm.isCreation).toBe(false);

    });

    test('is read only', () => {
        const wrapper = createWrapper({
            propsData:{
                independant: true,
                props: {
                    ...createForm(),
                    authorizations: {
                        create: false,
                        update: false
                    }
                }
            }
        });
        expect(wrapper.vm.isReadOnly).toBe(true);

        wrapper.setData({
            authorizations: {
                create: true,
                update: false,
            }
        });
        expect(wrapper.vm.isReadOnly).toBe(false);

        wrapper.setProps({
            instanceId: '10',
        });
        expect(wrapper.vm.isReadOnly).toBe(true);

        wrapper.setData({
            authorizations: {
                create: true,
                update: true,
            }
        });
        expect(wrapper.vm.isReadOnly).toBe(false);
        
        wrapper.setProps({
            ignoreAuthorizations: true,
        });
        wrapper.setData({
            authorizations: {
                create: false,
                update: false,
            }
        });
        expect(wrapper.vm.isReadOnly).toBe(false);
    })

    test('synchronous', () => {
        const wrapper = createWrapper({
            propsData:{
                independant: true,
                props: createForm(),
            }
        });

        expect(wrapper.vm.synchronous).toBe(true);

        wrapper.setProps({ independant: false });

        expect(wrapper.vm.synchronous).toBe(false);
    });


    test('has errors', () => {
        const wrapper = createWrapper();
        expect(wrapper.vm.hasErrors).toBe(false);
        
        wrapper.setData({
            errors: {
                field: ['required']
            }
        });

        expect(wrapper.vm.hasErrors).toBe(true);

        wrapper.setData({
            errors: {
                field: { cleared:true },
            }
        });

        expect(wrapper.vm.hasErrors).toBe(false);
    });

    test('expose appropriate props to layout components', () => {
        const wrapper = createWrapper({
            propsData: {
                independant: true,
                props: createForm()
            }
        });

        let { tabbedLayout, columnsGrid, fieldLayout } = wrapper.vm.$refs;

        expect(tabbedLayout.$options.propsData).toEqual({
            layout: { tabs:[{ columns:[{ fields:[[{key:'title'}]] }] }] } // $form.layout
        });
        expect(columnsGrid.$options.propsData).toEqual({
            rows: [[{ fields:[[{key:'title'}]] }]] // [$form.layout.tabs[0].columns]
        });
        expect(fieldLayout.$options.propsData).toEqual({
            layout: [[{key:'title'}]], // $form.layout.tabs[0].columns[0].fields
            visible: { title: true } // $form.fieldVisible
        });
    });

    test('expose appropriate props to field', async () => {
        const wrapper = createWrapper({
            propsData:{
                independant: true,
                props: {
                    ...createForm({ localized: true }),
                    locales: ['fr', 'en']
                }
            }
        });
        await wrapper.vm.$nextTick();
        const field = wrapper.find({ ref:'field' }).vm;

        expect(field.$attrs).toMatchObject({
            'field-key': 'title',
            'field-layout': { key: 'title' },
            'update-data': wrapper.vm.updateData,
            'locale': 'fr',
            'error-identifier': 'title',
            'config-identifier': 'title',
            'update-visibility': wrapper.vm.updateVisibility,
            'context-data': {
                title: null
            },
            'context-fields': {
                title: {
                    type: 'text',
                    readOnly: true
                },
            },
        });

        wrapper.setData({
            authorizations: { create: true }
        });

        expect(field.$attrs['context-fields'].readOnly).toBeFalsy();
    });

    test('update data', () => {
        const wrapper = createWrapper({
            propsData:{
                independant: true,
                props: createForm(),
            }
        });
        wrapper.vm.fieldLocalizedValue = jest.fn(()=>'fieldLocalizedValue');
        expect(wrapper.vm.data.title).toBe(null);

        wrapper.vm.updateData('title', 'text');
        expect(wrapper.vm.fieldLocalizedValue).toHaveBeenCalledWith('title', 'text');
        expect(wrapper.vm.data.title).toBe('fieldLocalizedValue');
    });


    test('update visibility', async () => {
        const wrapper = createWrapper({
            propsData:{
                independant: true,
                props: createForm()
            }
        });

        expect(wrapper.vm.fieldVisible.title).toBe(true);

        wrapper.vm.updateVisibility('title', false);

        expect(wrapper.vm.fieldVisible.title).toBe(false);
    });

    test('mount', () => {
        const wrapper = createWrapper({
            propsData:{
                independant: true,
                props: {
                    ...createForm({ localized:true }),
                    authorizations: {
                        create: true,
                        update: false
                    },
                    locales: ['en', 'fr']
                }
            }
        });

        expect(wrapper.vm).toMatchObject({
            ...createForm({ localized:true }),
            authorizations: {
                create: true,
                update: false
            },
            locales: ['en', 'fr'],
            fieldLocale: {
                title: 'en'
            }
        });

        wrapper.vm.patchLayout = ()=>{};
        wrapper.vm.ready = false;
        wrapper.vm.mount({ fields:{}, locales:null });

        expect(wrapper.vm.fieldLocale).toEqual({ title:undefined });
    });

    test('mount async', async () => {
        const wrapper = createWrapper();

        await nextRequestFulfilled({
            status: 200,
            response: {
                ...createForm(),
                authorizations: {
                    create: true,
                    update: false
                }
            }
        });

        expect(wrapper.vm).toMatchObject({
            ...createForm(),
            authorizations: {
                create: true,
                update: false
            }
        });
    });

    test('handle 422', async () => {
        const wrapper = createWrapper();

        wrapper.vm.actionsBus.$emit('submit');

        await nextRequestFulfilled({
            status: 422,
            response: {
                errors: {
                    title: ['invalid']
                }
            }
        });

        expect(wrapper.vm.errors).toEqual({
            title: ['invalid']
        });
    });

    test('patch layout', async () => {
        const wrapper = createWrapper({
            propsData:{
                independant: true,
                props: {
                    ...createForm(),
                    layout: createLayout( [
                        [{ legend:'fieldset', fields: [[]] }],
                        [{ legend:'fieldset', fields: [[]] }]
                    ]),
                }
            }
        });

        let {
            tabs : [
                {
                    columns: [{ fields }]
                }
            ]
        } = wrapper.vm.layout;

        expect(typeof fields[0][0].id).toBe('string');
        expect(fields[0][0].id).not.toEqual(fields[1][0].id)
    });


    test('setup action bar correctly', async () => {
        const wrapper = createWrapper();

        let setupEmitted = jest.fn();
        wrapper.vm.actionsBus.$on('setup', setupEmitted);
        jest.spyOn(wrapper.vm, 'setupActionBar');

        await nextRequestFulfilled({
            status: 200,
            response: {
                ...createForm(),
                authorizations: {
                    create: false,
                    update: false
                }
            }
        });

        expect(wrapper.vm.setupActionBar).toHaveBeenCalledTimes(1);
        expect(setupEmitted).toHaveBeenCalledTimes(1);
        expect(setupEmitted.mock.calls[0][0]).toMatchObject({
            showSubmitButton: false,
            showDeleteButton: false,
            showBackButton: true,
            opType: 'create'
        });


        wrapper.vm.authorizations.create = true;

        wrapper.vm.setupActionBar();

        expect(setupEmitted).toHaveBeenCalledTimes(2);
        expect(setupEmitted.mock.calls[1][0]).toMatchObject({
            showSubmitButton: true,
            showDeleteButton: false,
            showBackButton: false,
            opType: 'create'
        });


        wrapper.setProps({
            instanceId: '10',
        });
        wrapper.vm.authorizations.delete = true;

        wrapper.vm.setupActionBar();

        expect(setupEmitted).toHaveBeenCalledTimes(3);
        expect(setupEmitted.mock.calls[2][0]).toMatchObject({
            showSubmitButton: false,
            showDeleteButton: true,
            showBackButton: true,
            opType: 'update'
        });

        wrapper.vm.authorizations.update = true;

        wrapper.vm.setupActionBar();

        expect(setupEmitted).toHaveBeenCalledTimes(4);
        expect(setupEmitted.mock.calls[3][0]).toMatchObject({
            showSubmitButton: true,
            showDeleteButton: true,
            showBackButton: false,
            opType: 'update'
        });

    });

    test('redirect to list', async () => {
        const wrapper = createWrapper();

        expect(wrapper.vm.listUrl).toEqual('/sharp/list/spaceship?restore-context=1');

        wrapper.vm.redirectToList = jest.fn();

        wrapper.vm.actionsBus.$emit('delete');

        await nextRequestFulfilled({
            status: 200
        }, 0);

        expect(wrapper.vm.redirectToList).toHaveBeenCalledTimes(1);

    });

    test('submit', async () => {
        const wrapper = createWrapper({
            propsData: {
                independant:true,
                props: createForm(),
            },
        });

        wrapper.vm.post = jest.fn(()=>Promise.resolve({ data: { ok: true } }));

        wrapper.vm.pendingJobs.push('upload');
        wrapper.vm.actionsBus.$emit('submit');

        await wait(10);

        expect(wrapper.vm.post).not.toHaveBeenCalled();


        let submittedEmmitted = jest.fn();
        wrapper.vm.pendingJobs = [];

        wrapper.vm.actionsBus.$emit('submit');

        await wait(10);

        expect(wrapper.vm.post).toHaveBeenCalledTimes(1);
    });

    test('dependant submit', async () => {
        const wrapper = createWrapper();

        wrapper.vm.post = jest.fn(()=>Promise.resolve({ data: { ok: true } }));
        wrapper.vm.handleError = jest.fn();

        wrapper.vm.actionsBus.$emit('submit');

        await wait(10);

        expect(wrapper.vm.post).toHaveBeenCalledTimes(1);
        expect(wrapper.vm.post.mock.calls[0][0]).toBeUndefined();
        expect(wrapper.vm.post.mock.calls[0][1]).toBeUndefined();


        expect(wrapper.vm.handleError).not.toHaveBeenCalled();

        wrapper.vm.post = jest.fn(()=>Promise.reject({ error: true }));

        wrapper.vm.actionsBus.$emit('submit');

        await wait(10);

        expect(wrapper.vm.handleError).toHaveBeenCalledTimes(1);
        expect(wrapper.vm.handleError).toHaveBeenCalledWith({ error: true });
    });

    test('delete', async ()=>{
        // refer 'redirect to list' test
    });

    test('cancel', async ()=>{
        // refer 'redirect to list' test
    });

    test('pending jobs', async ()=> {
        const wrapper = createWrapper();
        let updateActionsStateEmitted = jest.fn();

        wrapper.vm.actionsBus.$on('updateActionsState', updateActionsStateEmitted);

        wrapper.vm.actionsBus.$emit('setPendingJob', { key:'myUpload', origin:'upload' ,value:true });

        expect(wrapper.vm.pendingJobs).toEqual(['myUpload']);
        expect(updateActionsStateEmitted).toHaveBeenLastCalledWith({
            state: 'pending',
            modifier: 'upload'
        });

        wrapper.vm.actionsBus.$emit('setPendingJob', { key:'myUpload', origin:'upload' ,value:false });

        expect(wrapper.vm.pendingJobs).toEqual([]);
        expect(updateActionsStateEmitted).toHaveBeenLastCalledWith(null);
    });

    test('has localize mixin with right fieldsProps', async () => {
        const wrapper = await createWrapper();
        expect(wrapper.vm.$options._localizedForm).toBe('fields');
    });
});