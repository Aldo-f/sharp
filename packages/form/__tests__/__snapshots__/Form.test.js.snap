// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sharp-form can mount "independant" sharp-form 1`] = `
<div data-popover-boundary="" class="SharpForm">
	<!---->
	<div id="MOCKED_TABBED_LAYOUT">
		<div id="MOCKED_GRID">
			<div id="MOCKED_FIELDS_LAYOUT">
				<fielddisplay-stub field-key="title" context-fields="[object Object]" context-data="[object Object]" field-layout="[object Object]" read-only="true" error-identifier="title" config-identifier="title" root="" update-data="function () { [native code] }"
				    update-visibility="function () { [native code] }"></fielddisplay-stub>
			</div>
		</div>
	</div>
</div>
`;

exports[`sharp-form can mount sharp-form 1`] = `
<div data-popover-boundary="" class="SharpForm">
	<!---->
	<div id="MOCKED_TABBED_LAYOUT">
		<div id="MOCKED_GRID">
			<div id="MOCKED_FIELDS_LAYOUT">
				<fielddisplay-stub field-key="title" context-fields="[object Object]" context-data="[object Object]" field-layout="[object Object]" read-only="true" error-identifier="title" config-identifier="title" root="" update-data="function () { [native code] }"
				    update-visibility="function () { [native code] }"></fielddisplay-stub>
			</div>
		</div>
	</div>
</div>
`;

exports[`sharp-form can mount sharp-form with alert 1`] = `
<div data-popover-boundary="" class="SharpForm">
	<div role="alert" class="alert alert-danger SharpForm__alert">
		<div class="fw-bold">{{ form.validation_error.title }}</div>
		<div>{{ form.validation_error.description }}</div>
	</div>
	<div id="MOCKED_TABBED_LAYOUT">
		<div id="MOCKED_GRID">
			<div id="MOCKED_FIELDS_LAYOUT">
				<fielddisplay-stub field-key="title" context-fields="[object Object]" context-data="[object Object]" field-layout="[object Object]" read-only="true" error-identifier="title" config-identifier="title" root="" update-data="function () { [native code] }"
				    update-visibility="function () { [native code] }"></fielddisplay-stub>
			</div>
		</div>
	</div>
</div>
`;
